restapi/src/main/java/io/stargate/web/docsapi/service/JsonDocumentShredder.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.stargate.web.docsapi.service;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.stargate.web.docsapi.exception.ErrorCode;
+import io.stargate.web.docsapi.exception.ErrorCodeRuntimeException;
+import io.stargate.web.docsapi.service.query.DocsApiConstants;
+import io.stargate.web.docsapi.service.util.DocsApiUtils;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import javax.inject.Inject;
+
+public class JsonDocumentShredder {
+
+  private final DocsApiConfiguration config;
+  private final ObjectMapper objectMapper;
+
+  @Inject
+  public JsonDocumentShredder(DocsApiConfiguration config, ObjectMapper objectMapper) {
+    this.config = config;
+    this.objectMapper = objectMapper;
+  }
+
+  public List<JsonShreddedRow> shred(
+      String payload, String documentId, List<String> subDocumentPath, boolean numericBooleans) {
+    try {
+      JsonNode node = objectMapper.readTree(payload);
+      return shred(node, documentId, subDocumentPath, numericBooleans);
+    } catch (JsonProcessingException e) {
+      throw new ErrorCodeRuntimeException(
+          ErrorCode.DOCS_API_INVALID_JSON_VALUE,
+          "Malformed JSON object found during shredding.",
+          e);
+    }
+  }
+
+  public List<JsonShreddedRow> shred(
+      JsonNode node, String documentId, List<String> subDocumentPath, boolean numericBooleans) {
+    // check if this is a valid root node
+    if (subDocumentPath.isEmpty()) {
+      checkRoot(node);
+    }
+
+    // sub-paths escaped
+    List<String> subPathsEscaped =
+        subDocumentPath.stream()
+            .map(DocsApiUtils::convertEscapedCharacters)
+            .collect(Collectors.toList());
```

Comment:
@EricBorczuk maybe we should leave this to caller and not touch the subDocumentPath?

Reply:
I think it would be reasonable to do it higher up, yea

Reply:
Done, resolving.